### PR TITLE
Mod 3 Consultancy Project

### PR DESCRIPTION
**Pull request by juliet-e**
# Curriculum PR Template

### Does this PR close any issues?
Yes - #238 

### Description
- Pull over and condense consultancy pages from BE M3
- Update rubric (FE eyes, please help me!)
- Adapt for async evals
- tried to consolidate project descriptions and avoid bloat
- added inception deliverables to website for consistency

### What questions do you have/what do you want feedback on? (optional)
1. Did I capture the FE requirements properly on the index page?
2. Please review the FE rubric notes (see TO-DOs)
3. Should we require a component architecture diagram?
4. BE folks -- how do the stretch tech options sound? They are all roughly the same pointage from List of Glory, but don't involve a Rails FE. Note the change to just one stretch tech and only for BE

### CODE REVIEW

File: module3/projects/consultancy/index.md [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+4. Assume one single user - multiple users with log in and authentication is out of scope.
+5. Implement functionality beyond simply data display - there needs to be some way for the user to work with or manipulate the data (favoriting, searching, filtering, commenting, rating, etc)
+
+<section class="dropdown">
+### Back End Requirements
+
+* Thorough testing at both the unit and integration level.
+  * Test Coverage tool like SimpleCov is used
+  * Happy Path and Sad Path/Edge Case testing is included
+* One Stretch Tech option (listed below) is implemented
```
Comment (epintozzi) [inline module3/projects/consultancy/index.md:64]: I like this idea of limited stretch choices versus the List of Glory which, while it has excellent ideas, can feel really overwhelming. Perhaps LOG moves over to Mod 4 or is otherwise documented somewhere for post-graduation practice.

**Comment by epintozzi** [comment]
Overall I like the approach you took for cleaning up the Consultancy spec. It's still pretty dense (for good reason). This might be a good project to start with 15-20 minutes of solo read through time as part of the kick off.

File: module3/projects/consultancy/index.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+---
+title: Consultancy Project
+layout: page
+---
+
+## Project Description
+
+The goal of this project is to create a successful web application from a student-led project idea. Your team will create an app that will solve a real world problem, but no problem is too small! You don't need to build an app that will save the world. What about an application that will gamify learning test questions for the written driving exam? Or an application that tells you when to move your car to avoid parking tickets for street sweeping? 
+
+Building student-designed projects can help differentiate you as a candidate, but it often comes down more to *how* you build it than *what* you build. The learning goals of this project are to expand/deepen your understanding of your learning process & dynamics within a team - NOT to build the flashiest thing ever.
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:10]: Love this

File: module3/projects/consultancy/index.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+This project will focus heavily on Service-Oriented Architecture (SOA) and will require at least one back end service (likely in Rails) and one React application.
+
+## Professional Development Benefits
+
+You will be better prepared to hit the ground running in the workplace having worked on a larger team, in a much larger application base, with lots of moving pieces. Understanding the deeper impact of how a small change can have a cascading effect on other pieces of a system will be crucial.
+
+### Feedback from Previous students:
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:18]: Love this section!! This would give me immediate buy-in as a student

File: module3/projects/consultancy/index.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+* Pagination
+* [Security Vulnerability Scanner](https://github.com/presidentbeef/brakeman)
+* OpenAPI Documentation
+* [Circuit Breaker Design Pattern](https://makisushi.io/posts/asd)
+* Observability (i.e. New Relic, Skylight, Honey Badger, Rails Performance)
+
+</section>
+
+<section class="dropdown">
+### Front End
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:81]: Rename to "Front End Requirements" to match the BE dropdown above this

File: module3/projects/consultancy/index.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+* [Presentation](./presentation)
+* [Evaluation](./evaluation)
+
+## Technical Requirements
+
+You have a lot of freedom with this project, but there are a few technical requirements. Overall, your project must:
+
+1. Includes at least one Rails API and one ReactJS application.
+2. Use one external API, but it can be consumed on the FE or BE. 
+3. Be deployed
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:54]: Do we want the frontend deployed? We might want to include some helpful hints for that since deploying React apps can be a little more complex than just using GH Pages

File: module3/projects/consultancy/index.md [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+</section>
+
+<section class="dropdown">
+### Front End
+
+* Utilizes React Router
+* Includes at least one POST request to the BE
+* Robust Cypress testing
+<!-- TO-DO: Require intercepts with Cypress tests? -->
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:86]: Yes, Cypress testing is not complete without intercepts.

Comment (kaylagordon) [reply module3/projects/consultancy/index.md:86]: In previous projects like this one ([Stretch Tech](https://frontend.turing.edu/projects/module-3/stretch.html)), we've made it really clear what amount of interactivity is necessary to be considered a solid MVP. Like, they can't just show us data, but they need to actually be able to manipulate it. I think that having POST called out helps with that, but we could be more detailed.

Comment (kaylagordon) [reply module3/projects/consultancy/index.md:86]:
```
* The application should be a multi-page application using React Router
* The application must include at least one POST to the backend
* The application cannot simply be a display of data - there needs to be some way for the user to work with or manipulate the data (favoriting, searching, filtering, commenting, rating, etc)
* The must be robust Cypress testing, with all network requests properly intercepted and all user flows tested
```

Comment (kaylagordon) [reply module3/projects/consultancy/index.md:86]: Also, question: We usually include "Assume one single user - multiple users with log in and authentication is out of the scope of this project." for their Stretch Tech project. Do we want to include that here, too?

File: module3/projects/consultancy/ideation.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+* Which API(s) are you thinking of using? If there are no free APIs, how else could you acquire external data?
+* Are any of the possible BE stretch techs particularly well-suited for this project?
+```
+
+### APIs
+
+Your application **must make good use of at least one external dataset or API**. Some examples include:
+
+#### Government Data
```
Comment (kaylagordon) [inline module3/projects/consultancy/ideation.md:54]: I love that we include a list of examples here. Finding public APIs can feel daunting to them!

File: module3/projects/consultancy/ideation.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+- [https://github.com/public-apis/public-apis](https://github.com/public-apis/public-apis)
+- [https://api.data.gov/list-of-apis/](https://api.data.gov/list-of-apis/)
+
+However, the list is not limited to these. You can choose to integrate with a service of your choosing, as long as it is approved ahead of time.
+
+---
+
+## Some Project Ideas to get the Creative Process Started
+
+### Previous Student Pitches
```
Comment (kaylagordon) [inline module3/projects/consultancy/ideation.md:115]: Love this too! I think students get overwhelmed trying to think of ideas. It's great giving them examples of appropriate scope for this project.

File: module3/projects/consultancy/inception.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
+---
```
Comment (kaylagordon) [inline module3/projects/consultancy/inception.md:1]: This page is a little messy with a mix of call out boxes and drop downs. I think just giving it a once-over to streamline the formatting would be helpful.

File: module3/projects/consultancy/index.md [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+We'll also examine our end users on a deeper level to instill a strong sense of developer empathy. We will do this by focusing time on "user personas" to answer questions like "who is our user", "what age range are we targeting", "how tech-savvy must the user be to use our app" and so on. As part of this project, groups will also perform an equity analysis in order to interrogate the product they're building, and reflect on possible changes that could better prioritize inclusion and harm reduction. 
+
+## Phases
+
+* [Ideation](./ideation)
+* [Inception](./inception)
+* [Development](./development)
+* [Retro](./retro_guide)
+* [Presentation](./presentation)
+* [Evaluation](./evaluation)
```
Comment (kaylagordon) [inline module3/projects/consultancy/index.md:46]: Are these 3 files supposed to be empty?

**Review by kaylagordon — COMMENTED** [review]
Overall, I love the content of this spec and think that we're setting them up for some really cool projects. 

I do think it's a lot of files to keep track of and I might get overwhelmed as a student trying to figure out where I should be in the various processes to be "on pace" to finish on time. I called it out in the `inception` file, but I think creating some uniformity around drop downs vs call out boxes could go a long way. Even using numbers in `index.md` like:

1. [Ideation](./ideation)
2. [Inception](./inception)
3. [Development](./development)
...etc

And then changing the titles of those pages to include those numbers like `title: Consultancy - Step 1 - Ideation`

And then maybe creating like a basic calendar of like:

Day 1: Complete Step 1. Ideation
Days 2-3: Complete Step 2. Inception
...etc.

IDK if that's too much. Y'all have given this project so much that you know what trips students up. I'm anticipating a lot of "Are we on track?" or student rushing through the planning phases because they don't think they should spend that much time on them.

File: module3/projects/consultancy/development.md [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+---
+_[Back to Consultancy Home](./index)_ 
+
+We will be using a (somewhat) Agile process during the development of this project. This will include:
+
+* Working in short 3 or 4 day "Sprints"
+* Using GitHub
+* Using a project management tool such as Github Projects
+* Conducting Story Grooming meetings
+* Doing daily Stand Ups
```
[the remark this answers is not in the document]
Comment (juliet-e) [reply module3/projects/consultancy/development.md:14]: Yeah, I've heard similar discussions, and a lot of teams use "daily" or "dailies" which I like. I'll add a note about this too.

File: module3/projects/consultancy/inception.md [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    - MVP definition
+    - Wireframes
+    - Chosen external API
+    - Chosen BE stretch tech
+    <!-- - React Component Architecture DO WE WANT THIS? -->
+    - Database Schema
+4. User Stories in Project Board
+5. Equity Analysis
+6. Plan for sub-teams for first development sprint
+    - Team rotation is required after sprints
```
[the remark this answers is not in the document]
Comment (juliet-e) [reply module3/projects/consultancy/inception.md:29]: Agreed - I think it'll be even more important to have students work on both BE and FE in this project because they will be different frameworks... but I've definitely seen teams struggle to do rotations even when it's all in Rails. One alternative is to require students to write "vertical" user stories that are full features that include any BE updates, but I worry that this could be too difficult for them to manage without prior experience with full-stack user stories.

File: module3/projects/consultancy/inception.md (hunk 9 above)
Comment (juliet-e) [reply module3/projects/consultancy/inception.md:1]: I tried to do dropdowns for all the sections, but any ## headers looked weird with dropdowns, so I kept those as is and just put the specific components of each of those inception sections in a dropdown. Not sure how else to organize it so if you have ideas, I'm happy to try them out!

File: module3/projects/consultancy/index.md (hunk 10 above)
Comment (juliet-e) [reply module3/projects/consultancy/index.md:46]: No they should have contents. I see them on here but are they not showing up for you?

Comment (juliet-e) [reply module3/projects/consultancy/index.md:46]: Nevermind -- there was definitely some git weirdness on my end but they should be here now!

File: module3/projects/consultancy/index.md (hunk 5 above)
Comment (juliet-e) [reply module3/projects/consultancy/index.md:54]: Yes I think we should. What do you think about requiring deployment for their pair project in M3 so they have a bit more practice before consultancy?

Comment (juliet-e) [reply module3/projects/consultancy/index.md:54]: If that's too much, we can add some deployment guidelines here. Do any exist already on the current FE site?

File: module3/projects/consultancy/index.md (hunk 6 above)
Comment (juliet-e) [reply module3/projects/consultancy/index.md:86]: Yes! The "assume one single user" is listed in the main technical requirements page! I stole the language straight from the stretch tech spec

File: module3/projects/consultancy/inception.md (hunk 9 above)
Comment (juliet-e) [reply module3/projects/consultancy/inception.md:1]: Nevermind I missed your overall comment! Will try to play around with that. I think I'm also not intentionally making callout boxes but if I use a `##` header it's happening? Going to remove those

**Comment by juliet-e** [comment]
> Overall, I love the content of this spec and think that we're setting them up for some really cool projects.
> 
> I do think it's a lot of files to keep track of and I might get overwhelmed as a student trying to figure out where I should be in the various processes to be "on pace" to finish on time. I called it out in the `inception` file, but I think creating some uniformity around drop downs vs call out boxes could go a long way. Even using numbers in `index.md` like:
> 
>     1. [Ideation](./ideation)
> 
>     2. [Inception](./inception)
> 
>     3. [Development](./development)
>        ...etc
> 
> 
> And then changing the titles of those pages to include those numbers like `title: Consultancy - Step 1 - Ideation`
> 
> And then maybe creating like a basic calendar of like:
> 
> Day 1: Complete Step 1. Ideation Days 2-3: Complete Step 2. Inception ...etc.
> 
> IDK if that's too much. Y'all have given this project so much that you know what trips students up. I'm anticipating a lot of "Are we on track?" or student rushing through the planning phases because they don't think they should spend that much time on them.

I like the idea of having numbered "steps" on inception. Something to note is that our calendar usually has `Work Time - Inception` and then `Work Time - Development` so that students know to complete inception first. Students can move to development time earlier only if they turn in their inception deliverables to their PM. 

Ideation is linked just in the project pitch reminders on the calendar, so only people interested in pitching a project will have to do that, rather than whole teams. I'll add a note about that in the index. 

I'm going to add a note about the timeline on the inception page too. Hopefully with these changes it should feel less overwhelming!

File: module3/projects/consultancy/evaluation.md [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+**Meets Expectations:**
+<!-- TO-DO FE instructor opinions here would be very helpful! -->
+* Code follows DRY and SRP design
+* API consumption is refactored and encapsulated, whether implemented on the FE or BE
+* Back End Expectations
+  - Routes are RESTful (PMs have some discretion here)
+  - Database migrations are managed well, and schema is well-planned and executed
+  - Model data validations are utilized
+  - Most common errors are rescued and serialized in a friendly error response.
+* Front End Expectations
```
Comment (juliet-e) [inline module3/projects/consultancy/evaluation.md:72]: @kaylagordon What do you think about including these requirements?

File: module3/projects/consultancy/inception.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (epintozzi) [reply module3/projects/consultancy/inception.md:29]: I like the idea of vertical stories. They're going to be doing both ends of the stack for the final project/assessment, so it's not a bad idea to encourage a first attempt at developing a whole feature prior to the final.

File: module3/projects/consultancy/evaluation.md [hunk 14]
```diff
@@ -0,0 +1,102 @@
+---
+layout: page
+title: Consultancy - Evaluation
+type: project
+---
+
+_[Back to Consultancy Home](./index)_ 
+
+## Rubric
```
Comment (epintozzi) [inline module3/projects/consultancy/evaluation.md:9]: Do we plan to add "Approaching" and "Below" expectation categories? Or are they required to complete all the bullet points of Meets Expectations to pass making the other two categories irrelevant?

File: module3/projects/consultancy/inception.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (juliet-e) [reply module3/projects/consultancy/inception.md:29]: Cool - let's try vertical splitting stories! I'll change this line to `6. Feature and user story assignment for first sprint`

File: module3/projects/consultancy/evaluation.md (hunk 14 above)
Comment (juliet-e) [reply module3/projects/consultancy/evaluation.md:9]: We actually currently don't have those sections for consultancy nor the FE stretch tech project. It's a bit of a departure from the other projects, but I just kept it the same. I'm not opposed to adding them

Comment (juliet-e) [reply module3/projects/consultancy/evaluation.md:9]: Especially since they can wiggle their MVP for the first 2 iterations, I think it's not unreasonable to require all the categories to be a passing project. I'd love to hear more about how the FE grades stretch tech. 

Right now with BE, we grade consultancy projects as if all 4 categories were present and the approaching/below grades are sort of assumed.. which is not great I suppose

File: module3/projects/consultancy/index.md [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+  * "Incredibly helpful in interviewing. It was the biggest project on my resume, and was key to my early interviews. It was more important in my interview for my new role than my Mod 4 capstone because of the multi-app architecture. I could speak to decision-making more because I got to work on each piece. My new company said I reset their expectations for what a bootcamp grad could accomplish."
+  * "It came up in just about every interview I had."
+  * "People seemed most excited/interested about working on a larger team project. They asked lots of followup questions about how we handled project management, made decisions, etc. It was really helpful to know how SOA/multi-app systems worked."
+  * "They were surprised that we got exposure to SOA and speaking about microservices.
+  It helped to understand and speak to Abstraction and Encapsulation a lot more. Working o
+  n a larger team definitely helped me get started at my job which was a way bigger team t
+  han the 10-person project."
+  * "People seemed most excited/interested about working on a larger team project (10 p
```
Comment (hfaerber) [inline module3/projects/consultancy/index.md:29]: @juliet-e There is an space between `p eople` and `m ade` and `t han`

File: module3/projects/consultancy/index.md [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+* [Circuit Breaker Design Pattern](https://makisushi.io/posts/asd)
+* Observability (i.e. New Relic, Skylight, Honey Badger, Rails Performance)
+
+</section>
+
+<section class="dropdown">
+### Front End Requirements
+
+* Multi-page application utilizing React Router
+* Includes at least one POST request to the BE
```
Comment (hfaerber) [inline module3/projects/consultancy/index.md:84]: Do we need to specific that a GET request is also required?  Not leave anything to be wrongly assumed?

File: module3/projects/consultancy/index.md [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+* The application cannot simply be a display of data - there needs to be some way for the user to work with or manipulate the data (favoriting, searching, filtering, commenting, rating, etc)
+* Includes client-side validations and error-handling
```
Comment (hfaerber) [inline module3/projects/consultancy/index.md:86]: Can we make 2 small wording changes?
The application cannot simply be a display of data - there needs to be **at least one way** for the user to work with or manipulate the data (favoriting, searching, filtering, commenting, rating, etc)
Includes **thorough** client-side validations and error-handling **including error handling for network requests**

Comment (hfaerber) [reply module3/projects/consultancy/index.md]: done

File: module3/projects/consultancy/index.md (hunk 5 above)
Comment (hfaerber) [reply module3/projects/consultancy/index.md]: I added a note to the FE requirements about researching and choosing a deployment platform for the FE application.  Done.
